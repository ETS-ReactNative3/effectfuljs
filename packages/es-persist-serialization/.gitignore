index.js
node_modules
tmp-*
*.orig
*.log
*~
*.map
.DS_Store
.tern-port
err-log.json
