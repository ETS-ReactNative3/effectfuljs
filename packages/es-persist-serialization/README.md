# @effectful/es-persist-serialization

The library uses [@effectful/es-persist](https://github.com/awto/effectfuljs/tree/master/packages/serialization)
to store async function state for
[@effectful/es-persist](https://github.com/awto/effectfuljs/tree/master/packages/es-persist).

It re-exports both of the libraries, and `@effectful/serialization` is
imported as peer dependency.

## Usage

Install from npm:

```
$ npm install --save-dev @effectful/es
$ npm install --save @effectful/es-persist-serialization @effectful/serialization
```

In .babelrc:

```json
{
  "presets": ["@effectful/es-persist-serialization/transform"]
}

```

If there are other babel presets, add `"passPerPreset":true`.

## License

Distributed under the terms of the [The MIT License (MIT)](LICENSE). 

