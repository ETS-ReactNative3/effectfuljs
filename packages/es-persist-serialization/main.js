import * as S from "@effectful/serialization";
import * as R from "@effectful/es-persist";
export {
  generator,
  generatorFunction,
  iterator,
  iteratorM,
  forInIterator,
  contextSymbol,
  sendSymbol,
  awaitSymbol,
  context,
  async,
  asyncGenerator,
  lock,
  all,
  any,
  producer,
  current,
  idle,
  abort,
  managed
} from "@effectful/es-persist";

export {
  read,
  write,
  regConstructor,
  regOpaqueObject,
  bind,
  descriptorSymbol,
  parse,
  stringify,
  regDescriptor,
  regOpaquePrim
} from "@effectful/serialization";

R.constructors.forEach(v => S.regConstructor(v));
export const asyncFunction = R.makeAsyncFunctionConstructor(S.regConstructor);
export const asyncGeneratorFunction = R.makeAsyncGeneratorFunctionConstructor(
  S.regConstructor
);
