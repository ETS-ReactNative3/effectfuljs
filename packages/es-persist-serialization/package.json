{
  "name": "@effectful/es-persist-serialization",
  "version": "1.3.6",
  "description": "@effectful/es-persist with @effectful/serialization for storing state",
  "main": "index.js",
  "module": "main.js",
  "jsnext:main": "main.js",
  "scripts": {
    "test": "mocha",
    "build": "babel --config-file ../es/babel.cjs.json -o index.js main.js"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/awto/effectfuljs.git"
  },
  "keywords": [
    "effectful",
    "effectfuljs",
    "monad",
    "linq"
  ],
  "author": "Vitaliy Akimov <vitaliy.akimov@gmail.com>",
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/awto/effectfuljs/issues"
  },
  "homepage": "https://github.com/awto/effectfuljs#readme",
  "peerDependencies": {
    "@effectful/serialization": "^1.0.0"
  },
  "devDependencies": {
    "@babel/cli": "^7.0.0",
    "@effectful/serialization": "^1.3.6",
    "mocha": "^5.0.5"
  },
  "dependencies": {
    "@effectful/es-persist": "^1.3.1"
  },
  "gitHead": "d3b25d1148c0b8da0bb45a6a52b8b8cac3150ca5"
}
