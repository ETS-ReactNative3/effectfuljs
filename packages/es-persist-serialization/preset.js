module.exports = require("@effectful/es-persist/preset")
  .options({importRT:"@effectful/es-persist-serialization"})
