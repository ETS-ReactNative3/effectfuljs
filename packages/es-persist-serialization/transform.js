module.exports = require("@effectful/es-persist/transform").options({
  importRT: "@effectful/es-persist-serialization"
});
