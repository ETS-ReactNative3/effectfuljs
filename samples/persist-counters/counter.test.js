import * as R from "@effectful/es-persist-serialization";
import React from "react";
import * as Counter from "./counter";
import renderer from "react-test-renderer";

async function snapshots(input) {
  for await (const i of input) expect(R.write(i)).toMatchSnapshot();
}

const controlMenuSample = [
  { type: "MENU", value: <span>Menu</span> },
  { type: "VALUE", value: 10 },
  { type: "CONTROL", value: <span>Control</span> },
  { type: "FLUSH" },
  { type: "VALUE", value: 11 },
  { type: "FLUSH" }
];

test("counterState", async () => {
  expect.assertions(6);
  const s = R.producer();
  const iter = Counter.state(s);
  s.send({ type: "INCREMENT" });
  s.send({ type: "INCREMENT" });
  s.send({ type: "DUMMY" });
  s.send({ type: "DECREMENT" });
  s.stop();
  expect(await iter.next()).toEqual({
    done: false,
    value: { type: "VALUE", value: 0 }
  });
  expect(await iter.next()).toEqual({
    done: false,
    value: { type: "VALUE", value: 1 }
  });
  expect(await iter.next()).toEqual({
    done: false,
    value: { type: "VALUE", value: 2 }
  });
  expect(await iter.next()).toEqual({ done: false, value: { type: "DUMMY" } });
  expect(await iter.next()).toEqual({
    done: false,
    value: { type: "VALUE", value: 1 }
  });
  expect(await iter.next()).toEqual({ done: true, value: void 0 });
});

test("counterControl", async () => {
  expect.assertions(3);
  for await (const i of Counter.mainControl(controlMenuSample))
    if (i.type === "CONTROL")
      expect(renderer.create(i.value).toJSON()).toMatchSnapshot();
});

test("prepControl", async () => {
  expect.assertions(8);
  await snapshots(Counter.prepControl(controlMenuSample));
});

const itemsSample = [
  {
    type: "ITEM",
    key: 10,
    value: { type: "CONTROL", value: <span>Item 1</span> }
  },
  { type: "ITEM", key: 10, value: { type: "VALUE", value: 1 } },
  {
    type: "ITEM",
    key: 100,
    value: { type: "CONTROL", value: <span>Item 2</span> }
  },
  { type: "ITEM", key: 100, value: { type: "VALUE", value: 10 } },
  {
    type: "ITEM",
    key: 1,
    value: { type: "CONTROL", value: <span>Item 3</span> }
  },
  { type: "ITEM", key: 1, value: { type: "VALUE", value: 100 } },
  { type: "ITEM", key: 100, value: { type: "DELETE" } }
];

test("controlsList", async () => {
  expect.assertions(4);
  for await (const i of Counter.controlsList(itemsSample))
    if (i.type === "CONTROL")
      expect(renderer.create(i.value).toJSON()).toMatchSnapshot();
});

test("totals", async () => {
  expect.assertions(4);
  for await (const i of Counter.totals(itemsSample))
    if (i.type === "MENU_ITEM")
      expect(renderer.create(i.value).toJSON()).toMatchSnapshot();
});
