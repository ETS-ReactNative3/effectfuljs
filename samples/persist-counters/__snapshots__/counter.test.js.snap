// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`controlsList 1`] = `
<span>
  Item 1
</span>
`;

exports[`controlsList 2`] = `
Array [
  <span>
    Item 1
  </span>,
  <span>
    Item 2
  </span>,
]
`;

exports[`controlsList 3`] = `
Array [
  <span>
    Item 3
  </span>,
  <span>
    Item 1
  </span>,
  <span>
    Item 2
  </span>,
]
`;

exports[`controlsList 4`] = `
Array [
  <span>
    Item 3
  </span>,
  <span>
    Item 1
  </span>,
]
`;

exports[`counterControl 1`] = `
<span>
  Control
</span>
`;

exports[`counterControl 2`] = `
Array [
  <p>
    Clicked: 
    10
     times
    <button
      onClick={[Function]}
    >
      +
    </button>
    <button
      onClick={[Function]}
    >
      -
    </button>
    <span>
      Menu
    </span>
  </p>,
  <div
    style={
      Object {
        "paddingLeft": 10,
      }
    }
  >
    <span>
      Control
    </span>
  </div>,
]
`;

exports[`counterControl 3`] = `
Array [
  <p>
    Clicked: 
    11
     times
    <button
      onClick={[Function]}
    >
      +
    </button>
    <button
      onClick={[Function]}
    >
      -
    </button>
    <span>
      Menu
    </span>
  </p>,
  <div
    style={
      Object {
        "paddingLeft": 10,
      }
    }
  >
    <span>
      Control
    </span>
  </div>,
]
`;

exports[`prepControl 1`] = `
Object {
  "type": "MENU",
  "value": Object {
    "#data": Object {
      "children": Array [
        "Menu",
      ],
      "props": Object {
        "children": undefined,
      },
      "type": "span",
    },
    "#type": "ReactElement",
  },
}
`;

exports[`prepControl 2`] = `
Object {
  "type": "VALUE",
  "value": 10,
}
`;

exports[`prepControl 3`] = `
Object {
  "type": "CONTROL",
  "value": Object {
    "#data": Object {
      "children": Array [
        "Control",
      ],
      "props": Object {
        "children": undefined,
      },
      "type": "span",
    },
    "#type": "ReactElement",
  },
}
`;

exports[`prepControl 4`] = `
Object {
  "type": "COLLECTION",
  "value": Object {
    "CONTROL": Object {
      "#data": Object {
        "children": Array [
          "Control",
        ],
        "props": Object {
          "children": undefined,
        },
        "type": "span",
      },
      "#type": "ReactElement",
    },
    "MENU": Object {
      "#data": Object {
        "children": Array [
          "Menu",
        ],
        "props": Object {
          "children": undefined,
        },
        "type": "span",
      },
      "#type": "ReactElement",
    },
    "VALUE": 10,
  },
}
`;

exports[`prepControl 5`] = `
Object {
  "type": "FLUSH",
}
`;

exports[`prepControl 6`] = `
Object {
  "type": "VALUE",
  "value": 11,
}
`;

exports[`prepControl 7`] = `
Object {
  "type": "COLLECTION",
  "value": Object {
    "CONTROL": Object {
      "#data": Object {
        "children": Array [
          "Control",
        ],
        "props": Object {
          "children": undefined,
        },
        "type": "span",
      },
      "#type": "ReactElement",
    },
    "MENU": Object {
      "#data": Object {
        "children": Array [
          "Menu",
        ],
        "props": Object {
          "children": undefined,
        },
        "type": "span",
      },
      "#type": "ReactElement",
    },
    "VALUE": 11,
  },
}
`;

exports[`prepControl 8`] = `
Object {
  "type": "FLUSH",
}
`;

exports[`totals 1`] = `
<b>
  Total 
  1
</b>
`;

exports[`totals 2`] = `
<b>
  Total 
  11
</b>
`;

exports[`totals 3`] = `
<b>
  Total 
  111
</b>
`;

exports[`totals 4`] = `
<b>
  Total 
  101
</b>
`;
